### PR TITLE
Chown files when copying into chroot

### LINKED ISSUE #2552: File ownership not considered when building chroots
### Nomad version
Nomad v0.5.5

### Operating system and Environment details
Ubuntu 16.04 on AWS us-east-1 c3.xlarge

### Issue
Inside the buildChroot method, there are some file copy and directory creation operations,
which are not considering file/dir ownership, only file/dir permissions.

### Reproduction steps
1. Create file /etc/foo/bar.txt, with perms 0400, owned by user/group bob:bob
2. mkdir /etc/bar, with perms 0550, owned by bob:bob
2. Create and run any simple job using the 'exec' driver (so that buildChroot is called)
3a. (When `data_dir` is on a separate filesystem from /): Inspect copied filesystem in alloc dir.
See that /etc/foo/bar.txt has perms 0400 but is owned by root:root (expected bob:bob).
See that /etc/bar has perms 0550 but is owned by root:root (expected bob:bob)
3b. (When `data_dir` is on same filesystem as /): Inspect hardlinked filesystem in alloc dir.
Remember that dirs cannot be hardlinked, nomad has to create them.
See that /etc/bar has perms 0550 but is owned by root:root (expected bob:bob)

It would be nice to preserve owner:group on these items.  Thanks!

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #2552

Not needed when hardlinking. Only adds Linux support but other Unixes may
be easy.

### CODE REVIEW

File: client/allocdir/alloc_dir.go [hunk 1]
```diff
@@ -400,7 +402,13 @@ func fileCopy(src, dst string, perm os.FileMode) error {
 	defer dstFile.Close()
 
 	if _, err := io.Copy(dstFile, srcFile); err != nil {
-		return fmt.Errorf("Couldn't copy %v to %v: %v", src, dst, err)
+		return fmt.Errorf("Couldn't copy %q to %q: %v", src, dst, err)
+	}
+
+	if uid >= 0 && gid >= 0 {
```
Comment (jshuping) [inline client/allocdir/alloc_dir.go]: Do we want to say `||` here in case the perms are like `root:bob`?

File: client/allocdir/fs_unix.go [hunk 2]
```diff
@@ -82,17 +82,17 @@ func getGid(u *user.User) (int, error) {
...
 	// TODO 0.6 clean this up. This was needed because chroot creation fails
 	// when a process restarts.
 	if fileInfo, _ := os.Stat(dst); fileInfo != nil {
 		return nil
 	}
 	// Attempt to hardlink.
-	if err := os.Link(src, dst); err == nil {
-		return nil
-	}
+	//if err := os.Link(src, dst); err == nil {
```
Comment (jshuping) [inline client/allocdir/fs_unix.go]: Were you maybe just doing this for testing / can we keep hardlinking?

File: client/allocdir/alloc_dir.go (hunk 1 above)
Comment (schmichael) [inline client/allocdir/alloc_dir.go]: I want to make sure they're both both valid IDs. Even if it's `root:bob` that's uid=0, gid>0 which will pass this conditional.

File: client/allocdir/fs_linux.go [hunk 3]
```diff
@@ -88,3 +88,11 @@ func removeSecretDir(dir string) error {
 	}
 	return os.RemoveAll(dir)
 }
+
+func getOwner(fi os.FileInfo) (int, int) {
+	stat, ok := fi.Sys().(*syscall.Stat_t)
```
Comment (dadgar) [inline client/allocdir/fs_linux.go]: Is `syscall.Stat_t` not implemented on the other platforms we chroot in? We shouldn't leave the bug open on other platforms we chroot in

Comment (sean-) [inline client/allocdir/fs_linux.go]: https://godoc.org/golang.org/x/sys/unix#Stat_t

**Comment by jshuping** [comment]
> Not needed when hardlinking.

We tried a build with the existing patch, but even when hardlinking files, the code is creating the directories to contain these hardlinks, without honoring the ownership (you can't hardlink directories).

I think we need to check the ownership of the source directory somehwere around here: https://github.com/hashicorp/nomad/blob/master/client/allocdir/task_dir.go#L176

**Comment by jcua** [comment]
With this change, this allows ownership of directories to be preserved.
 
```
diff --git a/client/allocdir/task_dir.go b/client/allocdir/task_dir.go
index d7e5fcf..e58d21a 100644
--- a/client/allocdir/task_dir.go
+++ b/client/allocdir/task_dir.go
@@ -155,6 +155,7 @@ func (t *TaskDir) embedDirs(entries map[string]string) error {
                        continue
                }

+               uid, gid := getOwner(s)
                // Embedding a single file
                if !s.IsDir() {
                        if err := createDir(t.Dir, filepath.Dir(dest)); err != nil {
@@ -163,7 +164,6 @@ func (t *TaskDir) embedDirs(entries map[string]string) error {

                        // Copy the file.
                        taskEntry := filepath.Join(t.Dir, dest)
-                       uid, gid := getOwner(s)
                        if err := linkOrCopy(source, taskEntry, uid, gid, s.Mode().Perm()); err != nil {
                                return err
                        }
@@ -178,6 +178,12 @@ func (t *TaskDir) embedDirs(entries map[string]string) error {
                        return fmt.Errorf("Couldn't create destination directory %v: %v", destDir, err)
                }

+               if uid >= 0 && gid >= 0 {
+                       if err := os.Chown(destDir, uid, gid); err != nil {
+                               return fmt.Errorf("Couldn't change owner of directory %v: %v", destDir, err)
+                       }
+               }
+
                // Enumerate the files in source.
                dirEntries, err := ioutil.ReadDir(source)
                if err != nil {
```

**Comment by sean-** [comment]
This looks good except for the use of the `syscall` package.  Instead, please find compatible syscalls from the `golang.org/x/sys/unix` package, which should address nearly all of the portability concerns.  If you want to merge this with `syscall` instead, let me know and I'll follow up with a PR to use the aforementioned `unix` package instead.

**Comment by schmichael** [comment]
@dadgar Fixed! A quick peek reveals all Unixes we support have `Stat_t.{U,G}id` fields, so we can have a single `getOwner` implementation in `fs_unix.go`.

@jshuping Great catch. Fixed!

@sean- The only place we use `syscall` is to type assert the return value of `FileInfo.Sys()`. To switch to `sys/unix` I think we'd need to change a lot of our cross-platform `os.Foo()`. I hate our mess of `fs_<OS>.go` helper files, so any PRs to clean it up are welcome!

**Comment by dadgar** [comment]
LGTM

**Pull request merged**